### PR TITLE
Put the argocd application status to ks application

### PR DESCRIPTION
**Pull request by LinuxSuRen**
### What this PR does / why we need it:
/kind feature

This feature can help us to get the application status. See the following test results:

```yaml
spec:
  argoApp:
    destination:
      namespace: default
      server: https://kubernetes.default.svc
    project: default
    source:
      directory:
        jsonnet: {}
        recurse: true
      path: test
      repoURL: https://gitlab.com/LinuxSuRen1/argocd
      targetRevision: HEAD
    syncPolicy:
      automated:
        prune: true
status:
  argoApp: '{"conditions":[{"lastTransitionTime":"2022-03-11T07:47:54Z","message":"failed
    to sync cluster https://10.43.0.1:443: failed to load initial state of resource
    Namespace: namespaces is forbidden: User \"system:serviceaccount:default:argocd-application-controller\"
    cannot list resource \"namespaces\" in API group \"\" at the cluster scope","type":"ComparisonError"},{"lastTransitionTime":"2022-03-11T07:47:54Z","message":"failed
    to sync cluster https://10.43.0.1:443: failed to load initial state of resource
    Namespace: namespaces is forbidden: User \"system:serviceaccount:default:argocd-application-controller\"
    cannot list resource \"namespaces\" in API group \"\" at the cluster scope","type":"ComparisonError"}],"health":{"status":"Healthy"},"reconciledAt":"2022-03-11T07:47:54Z","sync":{"comparedTo":{"destination":{"namespace":"default","server":"https://kubernetes.default.svc"},"source":{"directory":{"jsonnet":{},"recurse":true},"path":"test","repoURL":"https://gitlab.com/LinuxSuRen1/argocd","targetRevision":"HEAD"}},"status":"Unknown"}}'
```

### Which issue(s) this PR fixes:

Fixes #

### Special notes for reviewers:

Please check the following list before waiting reviewers:

- [ ] Already committed the CRD files to [the Helm Chart](https://github.com/kubesphere-sigs/ks-devops-helm-chart/) if you created some new CRDs
- [ ] Already [added the permission](https://github.com/kubesphere/ks-installer/blob/9e063b085a0e43fdb3d0d9e3e7f4149146f14b9c/roles/ks-core/prepare/files/ks-init/role-templates.yaml) for the new API
- [ ] Already added the RBAC markers for the new controllers

### Does this PR introduce a user-facing change??

```release-note
None
```

### CODE REVIEW

File: pkg/api/gitops/v1alpha1/application.go [hunk 1]
```diff
@@ -305,7 +305,13 @@ type Application struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-	Spec ApplicationSpec `json:"spec,omitempty"`
+	Spec   ApplicationSpec   `json:"spec,omitempty"`
+	Status ApplicationStatus `json:"status,omitempty"`
```
Comment (JohnNiang) [inline pkg/api/gitops/v1alpha1/application.go:309]: How about adding the status fields as subresource? Please refer to https://book.kubebuilder.io/reference/generating-crd.html?highlight=//+kubebuilder:subresource:status#subresources

File: controllers/argocd/argocd-application-status-controller.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"kubesphere.io/devops/pkg/api/gitops/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//+kubebuilder:rbac:groups=gitops.kubesphere.io,resources=applications,verbs=get;update
+//+kubebuilder:rbac:groups=argoproj.io,resources=applications,verbs=get;list
```
Comment (JohnNiang) [inline controllers/argocd/argocd-application-status-controller.go:32]: Don't forget adding the rbac config if you accept subresource suggestion below:
```go
//+kubebuilder:rbac:groups=gitops.kubesphere.io,resources=applications/status,verbs=get;update;patch
```

File: pkg/api/gitops/v1alpha1/application.go (hunk 1 above)
Comment (JohnNiang) [reply pkg/api/gitops/v1alpha1/application.go:309]: BTW, this won't block this PR from merging by other reviewers.

File: controllers/argocd/argocd-application-status-controller.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		err = nil
+		return
+	}
+
+	var status map[string]interface{}
+	if status, _, err = unstructured.NestedMap(argoCDApp.Object, "status"); err == nil {
+		var statusData []byte
+		if statusData, err = json.Marshal(status); err == nil {
+			app.Status.ArgoApp = string(statusData)
+			err = r.Update(ctx, app.DeepCopy())
```
[the remark this answers is not in the document]
Comment (LinuxSuRen) [reply controllers/argocd/argocd-application-status-controller.go]: Good catch!

**Review by JohnNiang — CHANGES_REQUESTED** [review]
Hi @LinuxSuRen , please execute `make manifests` first after you add subresource.

**Review by JohnNiang — APPROVED** [review]
/lgtm

**Pull request merged**